Implement task progress indicator (and dialog) in the toolbar

btut opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->
Implements a background-task progress indicator in JabRefs toolbar as first discussed in https://github.com/JabRef/jabref/pull/6381#discussion_r420043521.
![Screenshot from 2020-05-10 14-08-47](https://user-images.githubusercontent.com/10479048/81502520-bd077900-92de-11ea-8272-ec8859272a92.png)

The indicator is located at the right-most position in the toolbar. Clicking it opens a pop-over that lists the background tasks with an icon, title, message and progress.

In order not to overwhelm the user with background tasks, only interesting ones are shown. To make a background task show, it needs to have the showToUser property set. For it to have a meaningful entry, one should set the title and message to appropriate values. If progress can be tracked, that should be done as well. If not, an indeterminate indicator / progress bar is shown.
To set the icon of the task, add a mapping from the task-title to an icon to the iconMap in BackgroundTask.java

<!-- 
- Go through the list below. If a task has been completed, mark it done by using `[x]`.
- Please don't remove any items, just leave them unchecked if they are not applicable.
-->

- [X] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [X] Screenshots added in PR description (for UI changes)
- [X] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

tobiasdiez commented on src/main/java/org/jabref/gui/StateManager.java:
```diff
@@ -112,4 +118,20 @@ public void setSearchQuery(SearchQuery searchQuery) {
+        this.backgroundTasks.add(backgroundTask);
+    }
+
+    public BooleanBinding anyTaskRunningBinding = Bindings.createBooleanBinding(
```
For the bindings to update, you need to add the list as a dependency (second argument of the `createXBinding` method). In the case of lists, however, its easier to use `EasyBind.combine`: https://github.com/TomasMikula/EasyBind#combine-list

btut replied:
Aha, that looks useful.
However, I think I am running into some typing issues which I am struggling to resolve.

With the library you pointed me to, I wound up with the following:

`
    public Binding<Boolean> anyTaskRunningBinding = EasyBind.combine(
            backgroundTasks,
            stream -> stream.anyMatch(Task::getProgress)
    );
`

This gives me an error that it expects a Binding<Boolean>, but gets a MonadicBinding<R>.
If I just cast it, I get the following:

`no instance(s) of type variable(s) T exist so that Task conforms to ObservableValue<? extends T>`

I think this is the issue I need to resolve first. As the example in the library works fine, I guess the conversion from MonadicBinding<R> to Binding<Boolean> is then done implicitly, is that correct?

I struggle solving this because I dont know the type parameter of the tasks I am storing.
When storing, the Task has type V:

https://github.com/btut/jabref/blob/b8834914e5e735abaa7a710216888abe6bf54277/src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java#L99-L104

But in the list I just use Task:

https://github.com/btut/jabref/blob/b8834914e5e735abaa7a710216888abe6bf54277/src/main/java/org/jabref/gui/StateManager.java#L49

How do I need to change the list in order for it to work?

Siedlerchr replied:
Add the generics to the Task as welll:
`    private final UiThreadObservableList<Task<V>> backgroundTasks = new UiThreadObservableList(FXCollections.observableArrayList());`

btut replied:
But where would I get the V from in this case?
`Cannot resolve symbol 'V'`

btut replied:
I changed it to Task<?> now and now the progress, title and message properties make their way through to the dialogue.
I still cannot create the bindings for the progress indicator though.

btut replied:
That change also allows me to use EasyBind's listBind to bind the task list in the view to the task list in StateManager.

btut replied:
![image](https://user-images.githubusercontent.com/10479048/81391853-14aab680-911e-11ea-98a6-b63c1935a3df.png)
Here is the first peak at a download in the dialogue (which will probably be changed to a popover later). As you can see, the download task has it's title and message set and progress is updated fine. However, there are still a lot of tasks that do not have any details.

Siedlerchr replied:
Looks good already!
There are a lot of tasks which don't concern file downloads and just perform some operations in the background. If you search for all references from backgrond task you probably have to adjust each one to add a meaninful property

btut replied:
I think that should be doable. But before I do that: is it a good idea to have all background tasks listed? Or should we just show downloads? If we go with the first option, I suggest turning off the retain-tasks feature of the task view. Then, all completed tasks will automatically disappear. This way we have less tasks that only ran for a very little time filling up the view.

I got the bindings to work by storing a list of Property<Task<?>> Instead of Task<?>. That does the trick for the progress indicator. It now is indeterminate when one of the tasks has an indeterminate progress and shows the average progress otherwise (100% if no tasks are running).
For some reason, this breaks the task view. Since I now store a list of properties, and not a list of tasks, I cannot bind them directly to the view, so I went back to doing it manually, but that does not seem to work.

https://github.com/btut/jabref/blob/38dd89dce28b72d195de18b848b011532ef1f868/src/main/java/org/jabref/gui/taskprogressmanager/TaskProgressDialog.java#L34-L47

btut replied:
Now I just converted the list of properties back into a list of tasks using EasyBind and the bind that list to the list of tasks in the view (again using EasyBind). Now both the indicator and the dialogue work fine and are updated with the running tasks!

tobiasdiez commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -921,6 +930,24 @@ private MenuBar createMenu() {
+        indicator.setOnMouseClicked(new EventHandler<MouseEvent>() {
+            @Override
+            public void handle(MouseEvent event) {
+                TaskProgressDialog taskProgressDialog = new TaskProgressDialog();
```
Instead of using a real dialog, what about using a collapse overlay similar to how it's done in firefox?
https://github.com/controlsfx/controlsfx/wiki/ControlsFX-Features#popover

btut replied:
Uh fancy! I'll look into it!

btut replied:
![Screenshot from 2020-05-08 15-12-50](https://user-images.githubusercontent.com/10479048/81409085-c1952b80-913e-11ea-9935-36c17d30931d.png)
That looks much better! Great idea!

btut commented:
I would like to start a discussion here.

Right now, both the dialogue and the view are working (not perfect and pretty, but working).
At the very least, we have to manage the size of the vie wetter, as right now the progressbars are larger than the window so there is a horizontal scroll bar.
Also, we have to give all background tasks appropriate titles and messages for the view to be meaningful.

Also, there are many background tasks even after just starting up JabRef:
![image](https://user-images.githubusercontent.com/10479048/81402255-54c76480-9131-11ea-806c-4b082f7278bc.png)
I think this is a little overwhelming.
I suggest two options and am open to any ideas:

- Leave it as it is
- Show only tasks that are currently running. This will hide most startup tasks and the user will never see most of the tasks as they are too quick to complete.

Siedlerchr commented:
I would only add some tasks. Most tasks which are run in the background are irrelevant to the user

btut commented:
I changed it to only show download tasks. This looks much better!
![Screenshot from 2020-05-08 16-14-59](https://user-images.githubusercontent.com/10479048/81414433-45ebac80-9147-11ea-9936-615dd4528ce1.png)
I also changed the message to display the cite key instead of the url. The long url made the label in the task progress view so wide that it needed a scrollbar. Like this, it fits.

As you can see the progress indicator broke again, but I guess it never really displayed the correct value, the other tasks make me think that it did.

So next up I will fix that and then I will introduce a dialog that warns the user in case JabRef is closed during an ongoing download.

calixtus commented:
I would say: display all tasks that have a title (or put a property in the task: visible). So in theory, you could also add another button "show all" (or something similar).

calixtus commented:
Btw. I think this is a huge improvement to JabRef. Thank you for your effort.

btut commented:
> I would say: display all tasks that have a title (or put a property in the task: visible). So in theory, you could also add another button "show all" (or something similar).

I went with a property showToUser, a button to show all tasks does sound interesting, but I think that one only makes sense if all tasks set the message and title properties. When looking at the task-list, I guess that might be many pieces of code to adapt. Maybe it is better not to do that at first and maybe add that functionality later.

btut commented:
> Btw. I think this is a huge improvement to JabRef. Thank you for your effort.

Glad to be of help. It's a welcome distraction from my master's thesis :)

calixtus commented:
> Glad to be of help. It's a welcome distraction from my master's thesis :)

Yeah. This is exactly how I came (back) to JabRef and JabRef programming. 😆

tobiasdiez commented on src/main/java/org/jabref/gui/StateManager.java:
```diff
@@ -41,6 +48,7 @@
     private final OptionalObjectProperty<SearchQuery> activeSearchQuery = OptionalObjectProperty.empty();
     private final ObservableMap<BibDatabaseContext, IntegerProperty> searchResultMap = FXCollections.observableHashMap();
     private final OptionalObjectProperty<Node> focusOwner = OptionalObjectProperty.empty();
+    private final UiThreadObservableList<ObjectProperty<Task<?>>> backgroundTasks = new UiThreadObservableList(FXCollections.observableArrayList());
```
What's the reason to wrap the tasks around in a `ObjectProperty`? It should also work without this wrapper. If there were problems with the updates, you might need to add an "extractor" to the `observableArrayList` which specifies that the list should update if the underlying data changes (in this case probably the progress of the task).

btut replied:
I'm confused. I though EasyBind's combineList takes care of updating the observable when the progress updates? From the doc:

> Turns an observable list of observable values into a single observable value. The resulting observable value is updated when elements are added or removed to or from the list, as well as when element values change.

I had to turn it into a list of ObjectProperty because that's what EasyBind's combineList expects, but I guess it then only registers changes of the ObjectProperty, not the task.

So to update upon progress changes, I need to tell the list that I am interested in the progress by defining an extractor. I found some code online working with extractors, so I think I should be able to implement that.

tobiasdiez replied:
Yes, combineList takes updates into account but for this the elements have to observable themselves. For example, if `ObservableList<ObservableDoubleValue>` is a list containing the `progress` property of each task.

btut replied:
Thanks for the hints. This works!
![Screenshot from 2020-05-09 19-37-23](https://user-images.githubusercontent.com/10479048/81480977-4f4f4480-922d-11ea-8f57-658fc11430ed.png)

btut replied:
About the ObjectProperty wrapping issue. I tried changing it back just to see what happens, and I still get the same error for the bindings:

> no instance(s) of type variable(s) T exist so that Task<?> conforms to ObservableValue<? extends T>

I think that the combine method only works on lists of observables.
If I create the binding with Bindings.createDoubleBinding, it does not update.

tobiasdiez replied:
So
```java
public Binding<Double> tasksProgressBinding = Bindings.createDoubleBinding(
            backgroundTasks.filter(Task::isRunning).mapToDouble(Task::getProgress).average().orElse(1), backgroundTasks);`
```
doesn't work?

btut replied:
Actually, it does (except for some minor mistakes in your code).
I did not put the list name (backgroundTasks) as a second argument. Does that mean that the binding would have worked ONLY on the extractor? So adding and removing does not update the value, but a change in the progress would have? Or why else do we need to pass the list as a second argument?

tobiasdiez replied:
The second argument specifies the observables that the function (the first argument) depends on. That is, every time these observables change, the function is called and the binding is updated with the new value. If you don't specify any observables in the second argument, then the binding is never updated.

btut replied:
Oh that makes sense! Thanks for explaining!

btut commented:
I have a few issues with displaying a confirm dialogue when quitting while background tasks are still active.

For this purpose, I duplicated the WaitForSaveFinishedDialog. Could it be that this dialogue is not working in the current master? The showProgressDialogAndWait method in JabRefDialogService does not actually wait. I fixed that in https://github.com/JabRef/jabref/pull/6443/commits/62d7c142724573bdc5e65b1187c72f37ee401d7d. Is that ok?

I don't know why, but even after the fix JabRef does not wait for the dialogue to exit.

tobiasdiez commented:
Strange. I couldn't find any mistakes/issues in your code. Did you try to debug where the problem lies?

btut commented:
> Strange. I couldn't find any mistakes/issues in your code. Did you try to debug where the problem lies?
I added some debug messages. The task inside the dialog starts up and the loop-body is executed once. That's it. I cannot figure out why the loop does not run until the thread is canceled.

tobiasdiez commented on src/main/java/org/jabref/gui/WaitForBackgroundtasksFinishedDialog.java:
```diff
@@ -0,0 +1,49 @@
+                @Override
+                protected Void call() throws Exception {
+                    System.out.println("THREAD STARTED");
+                    while (stateManager.anyTaskRunningBinding.getValue()) {
```
Just a wild guess, could it be that actually all tasks are finished, because the `TaskExectutor` is already shutdown?

btut replied:
I added some logs in JabRefMain's stop method, it is run after the dialogue is already gone.

tobiasdiez commented on src/main/java/org/jabref/gui/WaitForBackgroundtasksFinishedDialog.java:
```diff
@@ -0,0 +1,49 @@
+                    Localization.lang("Please wait..."),
+                    Localization.lang("Waiting for background tasks to finish") + "...",
+                    waitForBackgroundtasksFinished
+            );
```
Add a log statement here as well to make sure that the progress dialog does indeed wait?

btut replied:
I currently have a log message after the loop to see whether the isCancelled property or my tasks-running property caused the stop. I now also put a log message where you suggested, just after the wait. They are exectued in the correct order. So yes, the wait does work, but the task exits too early.

btut replied:
I now also get a warning about the binding:

> May 10, 2020 12:02:50 AM javafx.beans.binding.Bindings$1 computeValue
WARNING: Exception while evaluating binding
java.lang.IllegalStateException: Task must only be used from the FX Application Thread
        at javafx.graphics/javafx.concurrent.Task.checkThread(Task.java:1220)
        at javafx.graphics/javafx.concurrent.Task.isRunning(Task.java:987)
        at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.base/java.util.AbstractList$RandomAccessSpliterator.tryAdvance(AbstractList.java:706)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:528)
        at org.jabref/org.jabref.gui.StateManager.lambda$new$1(StateManager.java:132)
        at javafx.base/javafx.beans.binding.Bindings$1.computeValue(Bindings.java:157)
        at javafx.base/javafx.beans.binding.BooleanBinding.get(BooleanBinding.java:155)
        at org.jabref/org.jabref.gui.WaitForBackgroundtasksFinishedDialog$1.call(WaitForBackgroundtasksFinishedDialog.java:28)
        at org.jabref/org.jabref.gui.WaitForBackgroundtasksFinishedDialog$1.call(WaitForBackgroundtasksFinishedDialog.java:24)
        at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.lang.Thread.run(Thread.java:832)

I tried to run the task on the Javafx thread by using:

`DefaultTaskExecutor.runInJavaFXThread(waitForBackgroundtasksFinished);`

The loop is then run more often (I guess until the download is finished), but the gui is frozen because of the task.

btut replied:
The problem was that the extractor I added to the task list in StateManager did not include the isRunning property. It still did not run perfectly and was not pretty, therefore I moved to a custom dialog which works fine.

btut commented:
I changed to custom dialog for the close action. This way, we don't need to start an extra task but can use the anyTaskRunning property from StateManager directly. Also, we can show all the running tasks:
![image](https://user-images.githubusercontent.com/10479048/81498043-e61a1080-92c2-11ea-9216-5c2fb0756435.png)
This works! The dialog closes itself when all tasks complete. The user can also choose to quit JabRef while tasks are still running or cancel the close-action.

tobiasdiez commented:
Very nice! This looks awesome!
(Can the save files background task also be integrated in this new dialog? It's ok to leave this as a new PR).

btut commented:
Looks even better with some Icons:
![Screenshot from 2020-05-10 14-08-47](https://user-images.githubusercontent.com/10479048/81498910-e7017100-92c7-11ea-8fcd-5120dec42228.png)

btut commented:
> Very nice! This looks awesome!
> (Can the save files background task also be integrated in this new dialog? It's ok to leave this as a new PR).

I agree, it would make sense to have one dialog only. But I think a separate PR would be better here.

btut commented:
> > Very nice! This looks awesome!
> > (Can the save files background task also be integrated in this new dialog? It's ok to leave this as a new PR).
> 
> I agree, it would make sense to have one dialog only. But I think a separate PR would be better here.

Actually, I think this can be achieved very easyliy, so we could include it here. The dialog for the sa

btut commented:
Sorry, don't know why this was closed. I must have clicked close on accident.

btut commented:
> > > Very nice! This looks awesome!
> > > (Can the save files background task also be integrated in this new dialog? It's ok to leave this as a new PR).
> > 
> > 
> > I agree, it would make sense to have one dialog only. But I think a separate PR would be better here.
> 
> I changed my mind and implemented it, but I don't know how to test if it is working. How can I provoke the saving to take so long that I notice?

To test I just made isSaving() return true always. The save-process shows up and is waited for. But there is a problem: I have to move this to after the user is asked if unsaved changes should be saved. But: the download tasks actually alter the database, so the saving should occur only after they are finished. I think doing it the old way, popping two seperate dialogs, might be safer here.

tobiasdiez commented:
Good point, the two dialog version does indeed sounds safer. Could you please add a short comment to the code explaining the reasoning.

I really appreciate your initiative and that you think about all the possible edge cases, too. Thanks!

btut commented:
Ok, I think that should be enough for this PR.
There are still a few checkstyle tests failing though because of the import-order being wrong. I checked all imports that cause errors and they are all in order (alphabetically and separated as they should be). Are there any other rules I am unaware of?

btut commented:
Also, the progress-indicator in the toolbar is larger if it's in indeterminate state. We should set it's prefered width so that it has the same size in both states. I don't know what value to choose though. I cannot find where the height of the toolbar is set. Is this size inferred from the size of the icons in it?

Siedlerchr commented:
Regarding import order, just use intellij or eclipse function organize imports then it should be in right order (if you imported style)

btut commented:
> Regarding import order, just use intellij or eclipse function organize imports then it should be in right order (if you imported style)

That worked, thanks!

btut commented:
What about the resizing of the indicator? Can someone help me out?
If not I would like to submit this PR for review. The resizing is very short and is not that bad.

tobiasdiez commented:
The size of the toolbar icons is specified here:

You can style the intermediate state using `.progress-indicator:indeterminate`. Maybe setting something like `-fx-fill-height: true;` (to the progress-indictaor itself or to `.progress-indicator:indeterminate > .spinner` might be enough). That's the original style:
https://github.com/openjdk/jfx/blob/4ec163df90c88cc01b70d58831ad2abdb86c196d/modules/javafx.controls/src/main/resources/com/sun/javafx/scene/control/skin/modena/modena.css#L1201-L1231

btut commented:
> The size of the toolbar icons is specified here:
> 
> https://github.com/JabRef/jabref/blob/b7e8a1515f1adb62f5a0a5e814c5b11982d9e08c/src/main/java/org/jabref/gui/Base.css#L881

Just setting the pref-width to 1.7em makes the indicator very very small. Can I somehow convert that value to be usable as a width?
![image](https://user-images.githubusercontent.com/10479048/81534946-3481ea80-9369-11ea-8514-822cc115e51d.png)


> You can style the intermediate state using `.progress-indicator:indeterminate`. Maybe setting something like `-fx-fill-height: true;` (to the progress-indictaor itself or to `.progress-indicator:indeterminate > .spinner` might be enough). That's the original style:
> https://github.com/openjdk/jfx/blob/4ec163df90c88cc01b70d58831ad2abdb86c196d/modules/javafx.controls/src/main/resources/com/sun/javafx/scene/control/skin/modena/modena.css#L1201-L1231

That didn't help. I tried to set it to the indicator, the indeterminate spinner and the group the indicator is placed in.

I think the issue is the following: The default indicator contains a label showing the percentage of work done. The label cannot be removed via styling, but I think that information is not suitable for a toolbar, so I clipped it to a square to remove the label, putting the indicator in a Group.
In indeterminate state the spinner takes up all the height, making the whole indicator wider. As I clip it to a square of width x width, it is now also higher and that causes the issue.

The clipping is done in a listener on the width property. A quick fix I just tested is to check if the indicator is currently indeterminate. If not, set the pref-width to the current width. This is not the most elegant solution, but it works well.

Sorry for not including the implementation details of the indicator in my previous question, it might have misled you.

tobiasdiez commented on src/main/java/org/jabref/gui/Base.css:
```diff
@@ -387,6 +387,21 @@
     -fx-padding: -0.1em 0.5em 0.5em 0.5em;
 }
 
+.progress-indicator {
```
The progress indicator is also used at other places. Since this definition here is global, it changes also the style of these other indicators, is this desired? I would guess some of the changes are ok globally (e.g the color) but others like the padding or no percentage are probably specific to the toolbar, right?

btut replied:
Make sense, I didn't see it anywhere else. I'll see to it.

btut replied:
Sounded easy, somehow doesn't work. I am not familiar with javafx, can you spot any mistakes in https://github.com/JabRef/jabref/pull/6443/commits/fcb1d0cc0ec3af744485b4d9321e3a3215c532f1? I tried to copy the style of icon-button which also has multiple styles.
The indicator is invisible now and can only be seen by the changing background color when hovering over it. The area that is then seen is very tiny.

btut replied:
I knew you would spot that mistake immediately, thanks! Works like a charm.
https://github.com/JabRef/jabref/pull/6443/commits/e24c1418564a9304b9c5172573e9edc9d1036a54

tobiasdiez commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -921,6 +940,45 @@ private MenuBar createMenu() {
+    private Group createTaskIndicator() {
+        ProgressIndicator indicator = new ProgressIndicator();
+        indicator.getStyleClass().setAll("progress-indicator");
+        indicator.progressProperty().bind(stateManager.tasksProgressBinding);
```
Also add `Background Tasks` as tooltip (if supported)?

btut replied:
I actually wanted to do that already, but forgot.

btut replied:
https://github.com/JabRef/jabref/pull/6443/commits/cf10859542c790e170866698f57a6fc8a704f7df

tobiasdiez commented on src/main/java/org/jabref/gui/StateManager.java:
```diff
@@ -41,6 +45,17 @@
+        return new Observable[] {taskProperty.progressProperty(), taskProperty.runningProperty()};
+    });
+
+    public BooleanBinding anyTaskRunningBinding = Bindings.createBooleanBinding(
```
Please make this fields private and add a public getter method (otherwise code from the outside can change these bindings, which is not what we want). (Maybe also remove the suffix "Binding" from the name, but that's a matter of taste).

btut replied:
Sure thing!

btut replied:
https://github.com/JabRef/jabref/pull/6443/commits/d56138bbb629dcd5ffdad9d29d9c7b87d80ba7c2

tobiasdiez commented on src/main/java/org/jabref/gui/WaitForBackgroundtasksFinishedDialog.java:
```diff
@@ -0,0 +1,64 @@
+        DialogPane contentPane = new DialogPane();
+        contentPane.setContent(box);
+
+        FXDialog alert = new FXDialog(Alert.AlertType.NONE, Localization.lang("Please wait..."));
```
https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/gui/DialogService.java#L179 (or the other overload) cannot be used here? I think it would be slightly cleaner to make `WaitForBackgroundtasksFinishedDialog` a "real" dialog (i.e. inherit from `Dialog`) or even convert it to a proper fxml-based dialog similar to most dialogs.

btut replied:
I did use the custom dialogue at the beginning, but it gives no access to the actual javafx dialogue, meaning I cannot hide it on an event. The thing with the JabRef dialogues created by the show*AndWait is that they create the actual javafx dialogue in the method, so they can't be accessed later. I just found the createDialog method which would probably be the best to use here. It creates the dialog and takes care of some styling and such, and then returns the actual dialog which can then be shown and waited for. I'll do that!

btut replied:
Is this ok? I tried to follow how showProgressDialogAndWait is implemented.
https://github.com/JabRef/jabref/pull/6443/commits/ff9ce0068d6b80e131826a9138aab3b1f981b00e

tobiasdiez commented on src/main/java/org/jabref/gui/util/BackgroundTask.java:
```diff
@@ -27,14 +33,29 @@
  */
 public abstract class BackgroundTask<V> {
+
+    public static ImmutableMap<String, Node> iconMap = ImmutableMap.of(
```
Why do you prefer a map here instead of adding a new property `icon` to the BackgroundTask class (so that `downloadTask.setIcon(IconTheme.JabRefIcons.DOWNLOAD)` works)?

btut replied:
That's because the TaskProgressView works on a list of tasks. Tasks have no icon property, therefore one must provide a callback that gets an icon from a task.
If BackgroundTask were derived from Task, we could add a property to BackgroundTask. The callback could then check if we have a BackgroundTask at hand and if so, provide the icon stored in a property in BackgroundTask. But BackgroundTask is not derived from Task so this is not possible.
We need a key to map from a Task to an Icon.
I used the title of the task as a key property to map to the icon, so an immutable map seemed like the best way to go.

tobiasdiez replied:
Ok, I see. It's a bit unfortunate but your solution makes sense. I would propose to change the `Callback` below to a normal method (`public static Node getIcon(Task<?> task)`) and replace the `iconMap` by a normal switch statement.

btut replied:
Change to method is easy, but I cannot use a switch, because I would need a constant expression. As I use the title of the task, and there is localization on that, it is not constant.
https://github.com/JabRef/jabref/pull/6443/commits/396411a68cf3a2583cbe367b716d6195f0773cf2

tobiasdiez commented on src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java:
```diff
@@ -96,7 +97,11 @@ public static void runInJavaFXThread(Runnable runnable) {
-        return execute(getJavaFXTask(task));
+        Task<V> javafxTask = getJavaFXTask(task);
+        if (task.showToUser()) {
+            Globals.stateManager.addBackgroundTask(javafxTask);
```
We try to remove the `Globals` class. Is it possible to insert the `stateManager` via a constructor parameter?

btut replied:
Yes, that should be easy.

btut replied:
Fixed in https://github.com/JabRef/jabref/pull/6443/commits/2c9ccea66fef2bcc6e68d0fadfdf407b1a0b6262

btut commented:
Great! Thanks for the quick review! There is a reason though I passed Objects to getIcon, that's the signature needed for setGraphicFactory in TaskProgressView.

tobiasdiez commented:
Ok I see, can you then please revert my two last commits (stupid me...sorry actually wanted to help :()

btut commented:
> Ok I see, can you then please revert my two last commits (stupid me...sorry actually wanted to help :()

Happens to the best :)

I have a couple of questions that are kind of off-topic here, I hope you don't mind:

- Why is the deployment failing?
- Do .deb packages built from [current master](https://builds.jabref.org/master/) contain a repository where updates can be fetched via apt? Or does one have to regularly fetch the new deb's himself?
- If these are not updated, how fast du you think these changes will make it into the ubuntu repos?

Completely unrelated:
- I found a bug in JabRef and am working on a fix. Shall I create an issue anyway or just create a PR once I am done?

btut commented:
> Completely unrelated:
> 
>     * I found a bug in JabRef and am working on a fix. Shall I create an issue anyway or just create a PR once I am done?

I am already done and just create a PR. I hope that is fine with you.

tobiasdiez commented:
> Why is the deployment failing?

It's a problem with our build server (no space left). You can safely ignore it.

> Do .deb packages built from current master contain a repository where updates can be fetched via apt? Or does one have to regularly fetch the new deb's himself?

The easiest way is to manually download the deb file or use snap. 

> If these are not updated, how fast du you think these changes will make it into the ubuntu repos?

We had problems with the official ubuntu repo. This is why the version there is pretty old (3.8 if I remember correctly). If you are interested, feel free to revisit the problem - maybe now that we provide deb packages this might be easier. @koppor may tell you more about the issues. Refs https://github.com/JabRef/jabref/issues/1371

Siedlerchr commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -397,7 +402,22 @@ private void tearDownJabRef(List<String> filenames) {
+                    Localization.lang("Please wait..."),
+                    Localization.lang("Waiting for background tasks to finish. Quit anyway?"),
+                    stateManager
+            ).orElse(ButtonType.CANCEL) == ButtonType.YES)) {
```
This looks odd to me

btut replied:
Could you please elaborate?
If you mean the condition at the end, I just want to make sure that JabRef only exists if the user pressed yes. So if the optional is empty I just put ButtonType.CANCEL so the comparison to YES returns false.
Do you want me to change this or add a comment? Maybe it would be clearer to store the result in a variable and check for (result.isPresent() && result.get() == ButtonType.YES).

Siedlerchr replied:
Yes, I think this is more readable.

btut replied:
Done https://github.com/JabRef/jabref/pull/6443/commits/3db399741b0ce098cd930005870bc745cd719c70

btut commented:
> We had problems with the official ubuntu repo. This is why the version there is pretty old (3.8 if I remember correctly). If you are interested, feel free to revisit the problem - maybe now that we provide deb packages this might be easier. @koppor may tell you more about the issues. Refs #1371

The one I was using on Ubuntu 20.04 is 5.0.50001-1.

tobiasdiez commented:
Thanks again!

btut commented:
Thanks for all your help along the way and for JabRef!
